plugins/ipc/stipc.cpp

```diff
@@ -89,10 +93,54 @@ static std::string layer_to_string(uint32_t layer)
     return "none";
 }
 
+static const struct wlr_pointer_impl pointer_impl = {
+    .name = "virtual-pointer",
+};
+
+wlr_input_device *virtual_pointer_device_create(
+    wlr_virtual_pointer_v1 *virtual_pointer)
+{
+    wlr_pointer_init(&virtual_pointer->pointer, &pointer_impl,
+        "wlr_virtual_pointer_v1");
+
+    return &virtual_pointer->pointer.base;
+}
+
+static void led_update(wlr_keyboard *keyboard, uint32_t leds)
+{}
+
+static const struct wlr_keyboard_impl keyboard_impl = {
+    .name = "virtual-keyboard",
+    .led_update = led_update,
+};
+
+wlr_input_device *virtual_keyboard_device_create(
+    wlr_virtual_keyboard_v1 *virtual_keyboard)
+{
+    wlr_keyboard_init(&virtual_keyboard->keyboard, &keyboard_impl,
+        "wlr_virtual_keyboard_v1");
+
+    return &virtual_keyboard->keyboard.base;
+}
+
+static const struct wlr_touch_impl touch_impl = {
+    .name = "virtual-touch-device",
+};
+
+wlr_input_device *virtual_touch_device_create(wlr_touch *virtual_touch)
+{
+    wlr_touch_init(virtual_touch, &touch_impl, "wlr_virtual_touch_v1");
+
+    return &virtual_touch->base;
+}
+
 class headless_input_backend_t
 {
   public:
     wlr_backend *backend;
+    wlr_virtual_pointer_v1 virtual_pointer;
```

Comment:
What is the problem with creating a wlr_pointer here, instead of `wlr_virtual_pointer_v1`? The virtual_pointer is the object created by the virtual-pointer protocol. Here, we create generic input devices.